service/controller/v1/prometheus/filter.go

```diff
@@ -10,8 +10,18 @@ func FilterInvalidServices(services []v1.Service) []v1.Service {
 	filteredServices := []v1.Service{}
 
 	for _, service := range services {
-		if _, ok := service.ObjectMeta.Annotations[ClusterAnnotation]; !ok {
-			continue
+		{
+			isDeleted := service.GetDeletionTimestamp() != nil
+			if isDeleted {
+				continue
+			}
+		}
+
+		{
+			_, hasAnnotation := service.ObjectMeta.Annotations[ClusterAnnotation]
```

Comment:
It looks like this should be a label and we should watch only such services.

Reply:
You are right on the concept but I am not too much into the details here. I am not sure how future proof the controller here is and what it's future looks like. There is a lot to be refactored and improved. I do not touch all the topics now. Maybe @giantswarm/sig-monitoring can say something about the future ideas here.